source/extensions/quic_listeners/quiche/envoy_quic_client_stream.cc

```diff
@@ -123,14 +125,16 @@ void EnvoyQuicClientStream::switchStreamBlockState(bool should_block) {
 
 void EnvoyQuicClientStream::OnInitialHeadersComplete(bool fin, size_t frame_len,
                                                      const quic::QuicHeaderList& header_list) {
-  quic::QuicSpdyStream::OnInitialHeadersComplete(fin, frame_len, header_list);
   if (rst_sent()) {
     return;
   }
-  ASSERT(headers_decompressed());
+  quic::QuicSpdyStream::OnInitialHeadersComplete(fin, frame_len, header_list);
+  ASSERT(headers_decompressed() && !header_list.empty());
+
   response_decoder_->decodeHeaders(
-      quicHeadersToEnvoyHeaders<Http::ResponseHeaderMapImpl>(header_list), /*end_stream=*/fin);
-  if (fin) {
+      quicHeadersToEnvoyHeaders<Http::ResponseHeaderMapImpl>(header_list),
+      /*end_stream=*/sequencer()->IsClosed());
+  if (sequencer()->IsClosed()) {
```

Comment:
Mind add a comment for the use of `sequencer()->IsClosed()` instead of `fin`? It is not obvious to me.